### PR TITLE
Fix: preserve order of invoices in bulk print

### LINKED ISSUE #12340: [Invoices] Bulk printing of invoices not reflecting sort applied in Orders screen
## Description

A user contacted support as they use the Bulk Actions - Print Invoices tool as packing slips for their orders. Since Tuesday 4th April, the user noticed an issue with the generated PDF where sorting applied to Orders was not being reflected in the PDF file. The user usually sorts the Orders page by name alphabetically, then clicks Print Invoices, and the generated PDF is also sorted alphabetically. Since Tuesday, the generated PDF is sorted differently (see Possible Fix below). The user is a hub which had 88 orders over 130 pages in their latest order cycle, so sorting the printed PDF manually is not practical.

## Expected Behavior

Generated PDFs should reflect sorting applied on the Orders screen, in the same way that filters (e.g. order cycle/name) are reflected in the generated PDF.

## Actual Behaviour

Generated PDF is sorted in one way, irrespective of sorting applied on Orders screen. Sorting by date or name does not change the generated PDF.

## Steps to Reproduce

1. User selects enterprise (ID: 202005), order cycle (id: 215923) sorts the Orders screen by clicking a column heading e.g. name column
Screenshot 2024-04-04 at 15 58 22
2. User sets view to '100 per page' to include all orders in outputted PDF.
3. User clicks 'select all' checkbox, then bulk actions, print invoices. 
4. Generated PDF is sorted by name alphabetically. 

## Animated Gif/Screenshot

First order in list when sorted alphabetically:
Screenshot 2024-04-04 at 15 58 22

First order in generated PDF:
Screenshot 2024-04-04 at 16 20 53

## Workaround

I've tried adding dummy company name/legal address into enterprise settings, which didn't work.
I've also tested with other enterprises and seen the same issue, as well as using the date range filter vs. the order cycle filter.
I've also tested on new admin and seen the same issue.
Same issue when viewing 15 per page.

## Severity

bug-s2: a non-critical feature is broken, no workaround

## Your Environment

* Version used: 
* Browser name and version:
* Operating System and version (desktop or mobile):

## Possible Fix

The sorting of orders in the PDF appears to be via the id field within spree_orders. Running this query on UK prod:
`select * from spree_orders where distributor_id=202005 and order_cycle_id=215923 and state='complete' order by id asc`
Returns the orders in the same order they appear in the PDF.
PDF:
Screenshot 2024-04-04 at 16 20 53
SQL query:
Screenshot 2024-04-04 at 16 48 49

### PR DESCRIPTION
**Pull request by mkllnk**
#### What? Why?

- Closes #12340 

Preserving the sort order in invoices again and simplifying the code a little.

#### What should we test?

- Bulk print invoices.
- Sort orders in different ways and check the invoice order.
- Repeat with the new invoices feature and without.

#### Release notes

Changelog Category (reviewers may add a label for the release notes):

- [x] User facing changes
- [ ] API changes (V0, V1, DFC or Webhook)
- [ ] Technical changes only
- [ ] Feature toggled

The title of the pull request will be included in the release notes.

### CODE REVIEW

File: app/jobs/bulk_invoice_job.rb [hunk 1]
```diff
@@ -8,13 +8,9 @@ class BulkInvoiceJob < ApplicationJob
   def perform(order_ids, filepath, options = {})
     @options = options
 
-    orders = Spree::Order.where(id: order_ids)
+    # The `find` method returns records in the same order as the given ids.
+    orders = Spree::Order.find(order_ids)
 
-    if OpenFoodNetwork::FeatureToggle.enabled?(:invoices, current_user)
-      orders = orders.invoiceable
-    end
```
Comment (mkllnk) [inline app/jobs/bulk_invoice_job.rb]: The OrdersReflex triggering this job was already applying this filter regardless of the feature toggle.

File: app/jobs/bulk_invoice_job.rb [hunk 2]
```diff
@@ -7,9 +7,14 @@ class BulkInvoiceJob < ApplicationJob
...
     @options = options
-    orders = sorted_orders(order_ids)
-    orders.filter!(&:invoiceable?) if OpenFoodNetwork::FeatureToggle.enabled?(:invoices,
-                                                                              current_user)
+
+    orders = Spree::Order.where(id: order_ids)
+
+    if OpenFoodNetwork::FeatureToggle.enabled?(:invoices, current_user)
+      orders = orders.invoiceable
+    end
```
Comment (mkllnk) [inline app/jobs/bulk_invoice_job.rb]: This is actually done by the reflex already and will be removed in the next commit.

File: app/jobs/bulk_invoice_job.rb [hunk 3]
```diff
@@ -22,9 +27,9 @@ def perform(order_ids, filepath, options = {})
   private
 
   # Ensures the records are returned in the same order the ids were originally given in
-  def sorted_orders(order_ids)
-    orders_by_id = Spree::Order.where(id: order_ids).to_a.index_by(&:id)
-    order_ids.map { |id| orders_by_id[id.to_i] }
+  def sort_orders(orders, order_ids)
+    orders_by_id = orders.to_a.index_by(&:id)
+    order_ids.map { |id| orders_by_id[id.to_i] }.compact
   end
```
Comment (mkllnk) [inline app/jobs/bulk_invoice_job.rb]: This will be replaced in the next commit as well.

**Comment by mkllnk** [comment]
After identifying the breaking commit, it took me one hour to solve the problem. And then two more hours to clean up the code. It wasn't really necessary but hopefully it will avoid bugs in the future.

File: app/jobs/bulk_invoice_job.rb [hunk 4]
```diff
@@ -8,13 +8,9 @@ class BulkInvoiceJob < ApplicationJob
   def perform(order_ids, filepath, options = {})
     @options = options
 
-    orders = Spree::Order.where(id: order_ids)
+    # The `find` method returns records in the same order as the given ids.
```
Comment (dacook) [inline app/jobs/bulk_invoice_job.rb:11]: TIL

File: app/reflexes/admin/orders_reflex.rb [hunk 5]
```diff
@@ -124,9 +131,8 @@ def all_distributors_can_invoice?(orders)
       Enterprise.where(id: distributor_ids, abn: nil).empty?
     end
```
Comment (dacook) [inline app/reflexes/admin/orders_reflex.rb:132]: `±` I guess this method can be removed too ☝️

**Review by dacook — CHANGES_REQUESTED** [review]
Thanks for picking this up, and doing a great job on refactoring and optimising. 🥇

I just noticed an extra method to remove.

File: app/reflexes/admin/orders_reflex.rb (hunk 5 above)
Comment (mkllnk) [reply app/reflexes/admin/orders_reflex.rb]: Yes, of course. Thanks for picking that up. I added a commit for that now.

**Review by dacook — APPROVED** [review]

**Comment by drummer83** [comment]
Hi @mkllnk,
Thanks for this important fix!

## Before staging
I confirmed that the sorting is not reflected in the order of the invoices. :heavy_check_mark:

## After staging
### Old Admin && New Invoices
Sorting by:
- Total ASC, DESC
- Name ASC, DESC
- Email ASC, DESC
- Shipment state ASC, DESC
- Payment state ASC, DESC
- Order state ASC, DESC
- Order number ASC, DESC
- Completed at ASC, DESC
- Results split across multiple results pages
- More than 50 results listed on one results page

### Admin v3 && New Invoices
Sorting by:
- Total ASC, DESC
- Name ASC, DESC
- Email ASC, DESC
- Shipment state ASC, DESC
- Payment state ASC, DESC
- Order state ASC, DESC
- Order number ASC, DESC
- Completed at ASC, DESC
- Results split across multiple results pages
- More than 50 results listed on one results page

### Admin v3 && Old Invoices
Sorting by:
- Total ASC, DESC
- Name ASC, DESC
- Email ASC, DESC
- Shipment state ASC, DESC
- Payment state ASC, DESC
- Order state ASC, DESC
- Order number ASC, DESC
- Completed at ASC, DESC
- Results split across multiple results pages
- More than 50 results listed on one results page

## Results
It's all looking good, thanks again! :partying_face: 
Merging! :rocket:

**Pull request merged**